File: ios/WaiComputer/WaiComputer/Features/Dictation/DictationHistoryStore.swift
import Foundation
import os
import WaiComputerKit

private let log = Logger(subsystem: "is.waiwai.computer.app", category: "dictation-history")

struct DictationHistoryEntry: Identifiable, Codable {
    let id: UUID
    let timestamp: Date
    let rawText: String
    let cleanedText: String?
    let durationSeconds: Double
    let wordCount: Int

    init(rawText: String, cleanedText: String?, durationSeconds: Double) {
        self.id = UUID()
        self.timestamp = Date()
        self.rawText = rawText
        self.cleanedText = cleanedText
        self.durationSeconds = durationSeconds
        self.wordCount = (cleanedText ?? rawText)
            .split(separator: " ")
            .count
    }

    var displayText: String {
        cleanedText ?? rawText
    }
}

/// Server-synced dictation history. Ported verbatim from macOS (pure
/// Foundation + WaiComputerKit + FileManager). The local cache lives in
/// Application Support and mirrors the server via the shared
/// `listDictationEntries` / `createDictationEntry` / `deleteDictationEntry`
/// endpoints. The view layer is iOS-native; this store is identical.
@MainActor
final class DictationHistoryStore: ObservableObject {
    @Published private(set) var entries: [DictationHistoryEntry] = []

    private let fileURL: URL
    private let tombstonesURL: URL
    private var tombstones: Set<UUID> = []
    private var apiClient: APIClient?

    init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = appSupport.appendingPathComponent("WaiComputer", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.fileURL = dir.appendingPathComponent("dictation_history.json")
        self.tombstonesURL = dir.appendingPathComponent("dictation_history_tombstones.json")
        load()
        loadTombstones()
    }

    // MARK: - Sync wiring

    /// Attach the authenticated API client. Once attached, mutations
    /// fan out to the server and `hydrate()` can pull/merge.
    func attach(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func add(rawText: String, cleanedText: String?, durationSeconds: Double) {
        let entry = DictationHistoryEntry(
            rawText: rawText,
            cleanedText: cleanedText,
            durationSeconds: durationSeconds
        )
        entries.insert(entry, at: 0)
        save()
        log.info("Saved dictation: \(entry.wordCount) words, \(String(format: "%.1f", durationSeconds))s")

        if apiClient != nil {
            Task { await self.pushEntry(entry) }
        }
    }

    func delete(_ entry: DictationHistoryEntry) {
        entries.removeAll { $0.id == entry.id }
        save()

        guard apiClient != nil else { return }
        tombstones.insert(entry.id)
        saveTombstones()
        Task { await self.deleteEntryOnServer(entry.id) }
    }

    /// Wipe the LOCAL cache only — server data is untouched. Called from
    /// the logout flow; on next login `hydrate()` repopulates from server.
    func clearLocalCache() {
        entries.removeAll()
        tombstones.removeAll()
        save()
        saveTombstones()
    }

    /// User-initiated "Clear All History" — wipes BOTH local and server.
    /// Each server delete is best-effort; failures land in tombstones so the
    /// next hydrate retries.
    func deleteAll() async {
        let snapshot = entries
        entries.removeAll()
        save()

        for entry in snapshot {
            if apiClient != nil {
                tombstones.insert(entry.id)
            }
        }
        saveTombstones()

        for entry in snapshot {
            await deleteEntryOnServer(entry.id)
        }
    }

    /// Pull the server's view, merge with local, then push any local-only
    /// entries up and retry any tombstoned deletes. Safe to call repeatedly.
    func hydrate() async {
        guard let apiClient else { return }

        let serverEntries: [DictationEntryDTO]
        do {
            serverEntries = try await apiClient.listDictationEntries()
        } catch {
            log.error("Hydrate fetch failed: \(error.localizedDescription)")
            return
        }

        // Step 2 — replay tombstoned deletes against server. Drop the
        // tombstone only after the server confirms.
        for serverEntry in serverEntries where tombstones.contains(serverEntry.clientEntryID) {
            do {
                try await apiClient.deleteDictationEntry(clientEntryID: serverEntry.clientEntryID)
                tombstones.remove(serverEntry.clientEntryID)
            } catch {
                log.error("Hydrate tombstone replay failed: \(error.localizedDescription)")
            }
        }
        saveTombstones()

        // Step 3 — server entries we don't have locally (and are not tombstoned).
        let localIDs = Set(entries.map(\.id))
        let serverByID = Dictionary(uniqueKeysWithValues: serverEntries.map { ($0.clientEntryID, $0) })
        var newLocal: [DictationHistoryEntry] = []
        for (id, dto) in serverByID where !localIDs.contains(id) && !tombstones.contains(id) {
            newLocal.append(Self.makeLocalEntry(from: dto))
        }
        if !newLocal.isEmpty {
            entries.append(contentsOf: newLocal)
            entries.sort { $0.timestamp > $1.timestamp }
            save()
        }

        // Step 4 — push local entries the server doesn't yet have.
        let serverIDs = Set(serverByID.keys)
        for entry in entries where !serverIDs.contains(entry.id) {
            await pushEntry(entry)
        }
    }

    // MARK: - Stats

    var totalWords: Int {
        entries.reduce(0) { $0 + $1.wordCount }
    }

    var averageWPM: Int {
        let totalDuration = entries.reduce(0.0) { $0 + $1.durationSeconds }
        guard totalDuration > 0 else { return 0 }
        return Int(Double(totalWords) / (totalDuration / 60.0))
    }

    var streakDays: Int {
        guard !entries.isEmpty else { return 0 }
        let calendar = Calendar.current
        var streak = 1
        var currentDate = calendar.startOfDay(for: Date())

        let dictatedToday = entries.contains { calendar.isDate($0.timestamp, inSameDayAs: currentDate) }
        if !dictatedToday {
            currentDate = calendar.date(byAdding: .day, value: -1, to: currentDate)!
            let dictatedYesterday = entries.contains { calendar.isDate($0.timestamp, inSameDayAs: currentDate) }
            if !dictatedYesterday { return 0 }
        }

        while true {
            let previousDay = calendar.date(byAdding: .day, value: -1, to: currentDate)!
            let dictatedOnDay = entries.contains { calendar.isDate($0.timestamp, inSameDayAs: previousDay) }
            if dictatedOnDay {
                streak += 1
                currentDate = previousDay
            } else {
                break
            }
        }
        return streak
    }

    // MARK: - Sync helpers

    private static let iso8601: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private func pushEntry(_ entry: DictationHistoryEntry) async {
        guard let apiClient else { return }
        let request = CreateDictationEntryRequest(
            clientEntryID: entry.id,
            rawText: entry.rawText,
            cleanedText: entry.cleanedText,
            durationSeconds: entry.durationSeconds,
            wordCount: entry.wordCount,
            occurredAt: Self.iso8601.string(from: entry.timestamp)
        )
        do {
            _ = try await apiClient.createDictationEntry(request)
        } catch {
            log.error("Push entry failed (will retry on next hydrate): \(error.localizedDescription)")
        }
    }

    private func deleteEntryOnServer(_ id: UUID) async {
        guard let apiClient else { return }
        do {
            try await apiClient.deleteDictationEntry(clientEntryID: id)
            tombstones.remove(id)
            saveTombstones()
        } catch {
            log.error("Delete on server failed (tombstone remains): \(error.localizedDescription)")
        }
    }

    private static func makeLocalEntry(from dto: DictationEntryDTO) -> DictationHistoryEntry {
        DictationHistoryEntry(
            id: dto.clientEntryID,
            timestamp: dto.occurredAt,
            rawText: dto.rawText,
            cleanedText: dto.cleanedText,
            durationSeconds: dto.durationSeconds,
            wordCount: dto.wordCount
        )
    }

    // MARK: - Persistence

    private func load() {
        guard FileManager.default.fileExists(atPath: fileURL.path) else { return }
        do {
            let data = try Data(contentsOf: fileURL)
            entries = try JSONDecoder().decode([DictationHistoryEntry].self, from: data)
            log.info("Loaded \(self.entries.count) history entries")
        } catch {
            log.error("Failed to load history: \(error.localizedDescription)")
        }
    }

    private func save() {
        do {
            try FileManager.default.createDirectory(
                at: fileURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            let data = try JSONEncoder().encode(entries)
            try data.write(to: fileURL, options: .atomic)
        } catch {
            log.error("Failed to save history: \(error.localizedDescription)")
        }
    }

    private func loadTombstones() {
        guard FileManager.default.fileExists(atPath: tombstonesURL.path) else { return }
        do {
            let data = try Data(contentsOf: tombstonesURL)
            tombstones = Set(try JSONDecoder().decode([UUID].self, from: data))
        } catch {
            log.error("Failed to load tombstones: \(error.localizedDescription)")
        }
    }

    private func saveTombstones() {
        do {
            try FileManager.default.createDirectory(
                at: tombstonesURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            let data = try JSONEncoder().encode(Array(tombstones))
            try data.write(to: tombstonesURL, options: .atomic)
        } catch {
            log.error("Failed to save tombstones: \(error.localizedDescription)")
        }
    }
}

// Internal initializer used by hydrate() to mint a local entry from server DTO.
extension DictationHistoryEntry {
    init(
        id: UUID,
        timestamp: Date,
        rawText: String,
        cleanedText: String?,
        durationSeconds: Double,
        wordCount: Int
    ) {
        self.id = id
        self.timestamp = timestamp
        self.rawText = rawText
        self.cleanedText = cleanedText
        self.durationSeconds = durationSeconds
        self.wordCount = wordCount
    }
}

File: desktop/WaiComputer.Core/Dictation/DictationHistoryStore.cs
using System.Linq;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using WaiComputer.Core.Api;
using WaiComputer.Core.Api.Models;
using WaiComputer.Core.Time;

namespace WaiComputer.Core.Dictation;

/// <summary>One completed dictation. Ports the macOS <c>DictationHistoryEntry</c>.</summary>
public sealed record DictationHistoryEntry(
    Guid Id,
    DateTimeOffset Timestamp,
    string RawText,
    string? CleanedText,
    double DurationSeconds,
    int WordCount)
{
    public string DisplayText => CleanedText ?? RawText;

    /// <summary>Word count of (cleaned ?? raw), collapsing runs of whitespace (matches Mac split).</summary>
    public static int CountWords(string text) => text.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries).Length;
}

/// <summary>
/// Local-first dictation history cached via <see cref="IDictationLocalStore"/>
/// (survives logout) + best-effort backend sync (tombstoned deletes, hydrate
/// merge). Ports the macOS <c>DictationHistoryStore</c>, including the stats
/// (total words, average WPM, day streak). Remote failures are logged + retried
/// on hydrate — the local cache is the session source of truth.
/// </summary>
public sealed class DictationHistoryStore
{
    private const string EntriesKey = "dictation_history";
    private const string TombstonesKey = "dictation_history_tombstones";

    private readonly IDictationLocalStore _local;
    private readonly IApiClient _api;
    private readonly ISystemClock _clock;
    private readonly ILogger<DictationHistoryStore> _logger;
    private readonly object _gate = new();
    private List<DictationHistoryEntry> _entries = new();
    private HashSet<Guid> _tombstones = new();

    public DictationHistoryStore(IDictationLocalStore local, IApiClient api, ISystemClock clock, ILogger<DictationHistoryStore>? logger = null)
    {
        _local = local;
        _api = api;
        _clock = clock;
        _logger = logger ?? NullLogger<DictationHistoryStore>.Instance;
    }

    public IReadOnlyList<DictationHistoryEntry> Entries
    {
        get { lock (_gate) { return _entries.ToList(); } }
    }

    public int TotalWords
    {
        get { lock (_gate) { return _entries.Sum(e => e.WordCount); } }
    }

    public int AverageWpm
    {
        get
        {
            lock (_gate)
            {
                var totalDuration = _entries.Sum(e => e.DurationSeconds);
                if (totalDuration <= 0) return 0;
                return (int)(_entries.Sum(e => e.WordCount) / (totalDuration / 60.0));
            }
        }
    }

    /// <summary>Consecutive days (UTC) ending today/yesterday with at least one dictation.</summary>
    public int StreakDays
    {
        get
        {
            HashSet<DateOnly> days;
            DateOnly today;
            lock (_gate)
            {
                if (_entries.Count == 0) return 0;
                days = _entries.Select(e => DateOnly.FromDateTime(e.Timestamp.UtcDateTime)).ToHashSet();
                today = DateOnly.FromDateTime(_clock.UtcNow.UtcDateTime);
            }

            var current = today;
            if (!days.Contains(current))
            {
                current = current.AddDays(-1);
                if (!days.Contains(current)) return 0; // nothing today or yesterday
            }

            var streak = 1;
            while (days.Contains(current.AddDays(-1)))
            {
                streak++;
                current = current.AddDays(-1);
            }
            return streak;
        }
    }

    public async Task LoadAsync(CancellationToken ct)
    {
        var entries = await _local.ReadAsync<List<DictationHistoryEntry>>(EntriesKey, ct).ConfigureAwait(false) ?? new List<DictationHistoryEntry>();
        var tombs = await _local.ReadAsync<List<Guid>>(TombstonesKey, ct).ConfigureAwait(false) ?? new List<Guid>();
        lock (_gate) { _entries = entries; _tombstones = new HashSet<Guid>(tombs); }
    }

    public async Task AddAsync(string rawText, string? cleanedText, double durationSeconds, CancellationToken ct)
    {
        var entry = new DictationHistoryEntry(
            Guid.NewGuid(),
            _clock.UtcNow,
            rawText,
            cleanedText,
            durationSeconds,
            DictationHistoryEntry.CountWords(cleanedText ?? rawText));
        lock (_gate) { _entries.Insert(0, entry); }
        await SaveEntriesAsync(ct).ConfigureAwait(false);
        await PushEntryAsync(entry, ct).ConfigureAwait(false);
    }

    public async Task DeleteAsync(Guid id, CancellationToken ct)
    {
        lock (_gate) { _entries.RemoveAll(e => e.Id == id); _tombstones.Add(id); }
        await SaveEntriesAsync(ct).ConfigureAwait(false);
        await SaveTombstonesAsync(ct).ConfigureAwait(false);
        try
        {
            await _api.DeleteDictationEntryAsync(id, ct).ConfigureAwait(false);
            lock (_gate) { _tombstones.Remove(id); }
            await SaveTombstonesAsync(ct).ConfigureAwait(false);
        }
        catch (Exception ex) { _logger.LogWarning(ex, "Delete dictation entry failed; tombstone retained"); }
    }

    public async Task ClearLocalCacheAsync(CancellationToken ct)
    {
        lock (_gate) { _entries.Clear(); _tombstones.Clear(); }
        await SaveEntriesAsync(ct).ConfigureAwait(false);
        await SaveTombstonesAsync(ct).ConfigureAwait(false);
    }

    /// <summary>Pull server entries, replay tombstoned deletes, merge server-only, push local-only.</summary>
    public async Task HydrateAsync(CancellationToken ct)
    {
        IReadOnlyList<DictationEntryDto> server;
        try { server = await _api.ListDictationEntriesAsync(ct).ConfigureAwait(false); }
        catch (Exception ex) { _logger.LogWarning(ex, "Hydrate history fetch failed"); return; }

        HashSet<Guid> tombs;
        HashSet<Guid> localIds;
        lock (_gate) { tombs = new HashSet<Guid>(_tombstones); localIds = _entries.Select(e => e.Id).ToHashSet(); }

        foreach (var s in server.Where(s => tombs.Contains(s.ClientEntryId)))
        {
            try { await _api.DeleteDictationEntryAsync(s.ClientEntryId, ct).ConfigureAwait(false); lock (_gate) { _tombstones.Remove(s.ClientEntryId); } }
            catch (Exception ex) { _logger.LogWarning(ex, "Hydrate tombstone replay failed"); }
        }
        await SaveTombstonesAsync(ct).ConfigureAwait(false);

        var additions = server
            .Where(s => !localIds.Contains(s.ClientEntryId) && !tombs.Contains(s.ClientEntryId))
            .Select(s => new DictationHistoryEntry(s.ClientEntryId, s.OccurredAt, s.RawText, s.CleanedText, s.DurationSeconds, s.WordCount))
            .ToList();
        if (additions.Count > 0)
        {
            lock (_gate)
            {
                _entries.AddRange(additions);
                _entries = _entries.OrderByDescending(e => e.Timestamp).ToList();
            }
            await SaveEntriesAsync(ct).ConfigureAwait(false);
        }

        var serverIds = server.Select(s => s.ClientEntryId).ToHashSet();
        List<DictationHistoryEntry> localOnly;
        lock (_gate) { localOnly = _entries.Where(e => !serverIds.Contains(e.Id)).ToList(); }
        foreach (var entry in localOnly)
        {
            await PushEntryAsync(entry, ct).ConfigureAwait(false);
        }
    }

    private async Task PushEntryAsync(DictationHistoryEntry entry, CancellationToken ct)
    {
        try
        {
            await _api.CreateDictationEntryAsync(
                new CreateDictationEntryRequest(entry.Id, entry.RawText, entry.CleanedText, entry.DurationSeconds, entry.WordCount, entry.Timestamp), ct).ConfigureAwait(false);
        }
        catch (Exception ex) { _logger.LogWarning(ex, "Push dictation entry failed; will retry on next hydrate"); }
    }

    private Task SaveEntriesAsync(CancellationToken ct)
    {
        List<DictationHistoryEntry> snapshot;
        lock (_gate) { snapshot = _entries.ToList(); }
        return _local.WriteAsync(EntriesKey, snapshot, ct);
    }

    private Task SaveTombstonesAsync(CancellationToken ct)
    {
        List<Guid> snapshot;
        lock (_gate) { snapshot = _tombstones.ToList(); }
        return _local.WriteAsync(TombstonesKey, snapshot, ct);
    }
}
